libs/server-sdk/src/data_sources/streaming_data_source.cpp

```diff
@@ -0,0 +1,152 @@
+    config::shared::built::HttpProperties http_properties,
+    boost::asio::any_io_executor ioc,
+    IDataSourceUpdateSink& handler,
+    DataSourceStatusManager& status_manager,
+    Logger const& logger)
+    : exec_(std::move(ioc)),
+      logger_(logger),
+      status_manager_(status_manager),
+      data_source_handler_(
+          DataSourceEventHandler(handler, logger, status_manager_)),
+      http_config_(std::move(http_properties)),
+      streaming_config_(
+          std::get<config::shared::built::StreamingConfig<
+              config::shared::ServerSDK>>(data_source_config.method)),
+      streaming_endpoint_(endpoints.StreamingBaseUrl()) {}
+
+void StreamingDataSource::Start() {
+    status_manager_.SetState(DataSourceStatus::DataSourceState::kInitializing);
+
+    auto updated_url = network::AppendUrl(streaming_endpoint_,
+                                          streaming_config_.streaming_path);
+
+    // Bad URL, don't set the client. Start will then report the bad status.
+    if (!updated_url) {
+        LD_LOG(logger_, LogLevel::kError) << kCouldNotParseEndpoint;
+        status_manager_.SetState(
+            DataSourceStatus::DataSourceState::kOff,
+            DataSourceStatus::ErrorInfo::ErrorKind::kNetworkError,
+            kCouldNotParseEndpoint);
+        return;
+    }
+
+    auto uri_components = boost::urls::parse_uri(*updated_url);
+
+    // Unlikely that it could be parsed earlier, and it cannot be parsed now.
+    if (!uri_components) {
+        LD_LOG(logger_, LogLevel::kError) << kCouldNotParseEndpoint;
+        status_manager_.SetState(
+            DataSourceStatus::DataSourceState::kOff,
+            DataSourceStatus::ErrorInfo::ErrorKind::kNetworkError,
+            kCouldNotParseEndpoint);
+        return;
+    }
+
+    // TODO: Initial reconnect delay. sc-204393
+    boost::urls::url url = uri_components.value();
+
+    auto client_builder = launchdarkly::sse::Builder(exec_, url.buffer());
+
+    client_builder.method(boost::beast::http::verb::get);
+
+    // TODO: can the read timeout be shared with *all* http requests? Or should
+    // it have a default in defaults.hpp? This must be greater than the
+    // heartbeat interval of the streaming service.
+    client_builder.read_timeout(std::chrono::minutes(5));
+
+    client_builder.write_timeout(http_config_.WriteTimeout());
+
+    client_builder.connect_timeout(http_config_.ConnectTimeout());
+
+    for (auto const& header : http_config_.BaseHeaders()) {
+        client_builder.header(header.first, header.second);
+    }
+
+    // TODO: Handle proxy support. sc-204386
+
+    auto weak_self = weak_from_this();
+
+    client_builder.receiver([weak_self](launchdarkly::sse::Event const& event) {
+        if (auto self = weak_self.lock()) {
+            self->data_source_handler_.HandleMessage(event.type(),
+                                                     event.data());
+            // TODO: Use the result of handle message to restart the
```

Comment:
I am not sure how to handle this in the current pattern. I think we went with re-creating instead of restarting. So it will require extracting out the build of the client, then closing and re-creating the client.

Reply:
Another thought is have the callback return a bool/status that tells the eventsource to reconnect. Then we don't have to rebuild it and can keep the async stuff hidden there.

Reply:
That sounds more pleasant.